hub/src/server/revocations.js

```diff
@@ -0,0 +1,121 @@
+/* @flow */
+
+import LRUCache from 'lru-cache'
+import type { DriverModel } from './driverModel'
+import fetch from 'node-fetch'
+import logger from 'winston'
+import { Readable } from 'stream'
+
+const MAX_AUTH_FILE_BYTES = 1024
+const AUTH_TIMESTAMP_FILE_NAME = 'authTimestamp'
+
+export class AuthTimestampCache {
+
+  cache: LRUCache<string, number>
+  driver: DriverModel
+  currentCacheEvictions: number
+
+  constructor(driver: DriverModel, maxCacheSize: number) {
+    this.currentCacheEvictions = 0
+    this.cache = new LRUCache<string, number>({ 
+      max: maxCacheSize, 
+      dispose: () => {
+        this.currentCacheEvictions++
+      }
+    })
+    this.driver = driver
+
+    // Check cache evictions every 10 minutes
+    const tenMinutes = 1000 * 60 * 10
+    this.setupCacheEvictionLogger(tenMinutes)
+  }
+
+  setupCacheEvictionLogger(timerInterval: number) {
+    const evictionLogTimeout: any = setInterval(() => this.handleCacheEvictions(), timerInterval)
+    evictionLogTimeout.unref()
+  }
+
+  handleCacheEvictions() {
+    if (this.currentCacheEvictions > 0) {
+      logger.warn(`Gaia authentication token timestamp cache evicted ${this.currentCacheEvictions} entries in the last 10 minutes. Consider increasing 'authTimestampCacheSize'.`)
+      this.currentCacheEvictions = 0
+    }
+  }
+
+  getAuthTimestampFileDir(bucketAddress: string) {
+    return `${bucketAddress}-auth`
+  }
+
+  async readAuthTimestamp(bucketAddress: string): Promise<number> {
+
+    const readUrlPrefix = this.driver.getReadURLPrefix()
+    const authTimestampDir = this.getAuthTimestampFileDir(bucketAddress)
+    
+    // Check if an auth number file exists
+    // TODO: Is the error for "file not exists" consistent enough to depend upon instead of doing this?
+    const bucketFileList = await this.driver.listFiles(authTimestampDir, null)
```

Comment:
Why do we do a listFiles here? Why not just let the subsequent fetch throw an error, and return `0` ?

Reply:
Yeah, this shouldn't use `listFiles` if at all possible.  `listFiles` isn't guaranteed to be consistent or performant across all the storage systems we'd like to use.

Reply:
That's what I suspected. I'm performing the fetch below here, but wasn't confident about what to do in any given error context. What happens if it is just a http timeout but there is a revocation timestamp? Then returning 0 would allow any potentially revoked tokens to work.

Reply:
If there's an HTTP timeout in loading the revocation timestamp, and the token has revocation data (i.e. an `iat` field), then I think the entire request should fail with 403 -- the caller is clearly asking for the revocation logic to be applied, and the authentication code's inability to be certain about the validity of the token under these circumstances implies that the _safest_ default behavior should be taken.

Reply:
I think a `404` should result in a `0` here.

In all likelihood, clients will _always_ set an `iat` field when generating tokens in the future. In the case that they haven't called `/revoke-all/` yet, all of their calls will 403 (even calls to `revoke-all`!) I think that's clearly bad behavior. A `404` should only occur if the file really isn't there -- in the event of a malicious storage provider, none of this matters anyways, they can write data whenever they want regardless of token validity.

Reply:
This has been updated to no longer user `driver.listFiles()` and instead uses the fetch response - in accordance to what we discussed.